File: server/games/example/lib/npc_movement.rb
module Example
  class NpcMovement
    include GameMachine::Commands

    attr_accessor :updates_between_move
    attr_reader :current_target, :position, :updates_for_move,
      :position_changed, :has_target, :last_move, :id,
      :reached_target, :current_distance_to_target
    attr_accessor :speed_scale
    def initialize(id,position)
      @id = id
      @position = position
      @last_move = Time.now.to_f
      @position_changed = false
      @has_target = false
      @reached_target = false
      @updates_for_move = 0
      @updates_between_move = 15
      @speed_scale = 4
      @current_distance_to_target = 0
      commands.grid.track(id,position.x,position.y,position.z)
    end

    def update
      @position_changed = false

      if has_target && !reached_target
        if @updates_for_move >= updates_between_move
          move
          @updates_for_move = 0
        end
      else
        return
      end

      # Only update position if we moved
      if position_changed
        commands.grid.track(id,position.x,position.y,position.z)
        @position_changed = false
      end

      @updates_for_move += 1
    end

    def distance_to_target
      current_distance_to_target
    end

    def set_target(target)
      update_target(target)
      @last_move = Time.now.to_f
      @has_target = true
      @reached_target = false
    end

    def update_target(target)
      @current_distance_to_target = position.distance(target)
      @current_target = target

      if current_distance_to_target == 0
        @reached_target = true
        return
      end

      current_distance_to_target
    end

    def drop_target
      @current_target = nil
      @has_target = false
      @reached_target = false
      @current_distance_to_target = 0
    end

    def set_reached_target
      position.x = current_target.x
      position.y = current_target.y
      @reached_target = true
    end

    def move
      @current_distance_to_target =  position.distance(current_target)
      if current_distance_to_target == 0
        set_reached_target
        return
      end

      delta_time = Time.now.to_f - last_move.to_f

      # Save object creation by not using methods that return new vector
      x = current_target.x - position.x
      y = current_target.y - position.y
      dirx,diry = GameMachine::Vector.norm(x,y)
      position.x += dirx * speed_scale * delta_time
      position.y += diry * speed_scale * delta_time

      #dir = (current_target - position).normalize
      #position.x += dir.x * speed_scale * delta_time
      #position.y += dir.y * speed_scale * delta_time


      if position.distance(current_target) > current_distance_to_target
        set_reached_target
      end

      @last_move = Time.now.to_f
      @position_changed = true
    end
  end
end

File: agents/agent_library_example/src/main/java/user/NpcMovement.java
package user;

import com.game_machine.util.Vector3;

public class NpcMovement {

	private Vector3 target;
	private Vector3 position;
	private double distanceToTarget = 0;
	private double lastMove = 0;
	private double speedScale = 4f;
	public boolean hasTarget = false;
	public boolean reachedTarget = false;

	public NpcMovement(Vector3 position, double speedScale) {
		this.speedScale = speedScale;
		this.position = position;
		this.lastMove = System.currentTimeMillis();
	}

	public void update() {
		if (hasTarget && !reachedTarget) {
			move();
		} else {
			return;
		}
	}

	private void targetReached() {
		// System.out.println("Target reached "+target.toString());
		position.x = target.x;
		position.y = target.y;
		reachedTarget = true;
	}

	public void setTarget(Vector3 newTarget) {
		target = newTarget;
		lastMove = System.currentTimeMillis();
		hasTarget = true;
		reachedTarget = false;
	}

	private void move() {
		distanceToTarget = position.distance2d(target);
		if (distanceToTarget <= 1) {
			targetReached();
			return;
		}
		double deltaTime = (System.currentTimeMillis() - lastMove) / 1000f;

		Vector3 dist = Vector3.zero();
		dist.x = target.x - position.x;
		dist.y = target.y - position.y;
		dist.normalizeLocal();

		position.x += dist.x * speedScale * deltaTime;
		position.y += dist.y * speedScale * deltaTime;

		// we moved past the target
		if (position.distance2d(target) > distanceToTarget) {
			targetReached();
		}

		lastMove = System.currentTimeMillis();
	}

}
